diff --git a/Config.in b/Config.in
index 3db90a6..d4c79e4 100644
--- a/Config.in
+++ b/Config.in
@@ -50,7 +50,7 @@ endchoice
 config CONFIG_IMG_ADDRESS
 	string
 	default "0x00008000" if CONFIG_FLASH
-	default "0x00008400" if CONFIG_DATAFLASH
+	default "0x00004000" if CONFIG_DATAFLASH 
 	default "0x00020000" if CONFIG_NANDFLASH
 	default	"0x00000000" if CONFIG_SDCARD
 	help
diff --git a/board/Config.in b/board/Config.in
index 3e8f1aa..af34af4 100644
--- a/board/Config.in
+++ b/board/Config.in
@@ -57,6 +57,18 @@ config	CONFIG_AT91SAM9263EK
 	help
 	  Use the AT91SAM9263EK Development board
 
+config	CONFIG_TNY_A9263
+	bool "tny-a9263"
+	select CONFIG_SDRAM
+	select ALLOW_NANDFLASH
+	select ALLOW_CPU_CLK_240MHZ
+	select ALLOW_CPU_CLK_200MHZ
+	select ALLOW_CPU_CLK_180MHZ
+	select ALLOW_CRYSTAL_12_000MHZ
+	select ALLOW_NANDFLASH_RECOVERY
+	help
+	  Use the TNY-A9263 Embedded Computer (silicon REVB Boot from nand supported)
+
 config	CONFIG_AT91SAM9RLEK
 	bool "at91sam9rlek"
 	select ALLOW_DATAFLASH
@@ -199,6 +211,7 @@ config CONFIG_CHIP
 	default "AT91SAM9260"	if CONFIG_AT91SAM9260EK
 	default "AT91SAM9261"	if CONFIG_AT91SAM9261EK
 	default "AT91SAM9263"	if CONFIG_AT91SAM9263EK
+	default "AT91SAM9263"	if CONFIG_TNY_A9263
 	default "AT91SAM9RL"	if CONFIG_AT91SAM9RLEK
 	default "AT91SAM9XE"	if CONFIG_AT91SAM9XEEK
 	default "AT91SAM9G10"	if CONFIG_AT91SAM9G10EK
@@ -218,6 +231,7 @@ config CONFIG_BOARD
 	default "at91sam9260ek"	if CONFIG_AT91SAM9260EK
 	default "at91sam9261ek"	if CONFIG_AT91SAM9261EK
 	default "at91sam9263ek"	if CONFIG_AT91SAM9263EK
+	default "tny-a9263"	if CONFIG_TNY_A9263
 	default "at91sam9rlek"	if CONFIG_AT91SAM9RLEK
 	default "at91sam9xeek"	if CONFIG_AT91SAM9XEEK
 	default "at91sam9g10ek"	if CONFIG_AT91SAM9G10EK
@@ -237,6 +251,7 @@ config CONFIG_MACH_TYPE
 	default "0x44B"	if CONFIG_AT91SAM9260EK
 	default "0x350"	if CONFIG_AT91SAM9261EK
 	default "0x4B2"	if CONFIG_AT91SAM9263EK	
+	default "0x85C" if CONFIG_TNY_A9263
 	default "1326"	if CONFIG_AT91SAM9RLEK
 	default "0x44B"	if CONFIG_AT91SAM9XEEK
 	default "0x86F"	if CONFIG_AT91SAM9G10EK		
@@ -271,10 +286,12 @@ config CONFIG_TOP_OF_MEMORY
 	default "0x301000"	if CONFIG_AT91SAM9260EK
 	default "0x328000"	if CONFIG_AT91SAM9261EK
 	default "0x314000"	if CONFIG_AT91SAM9263EK	
+	default "0x314000"	if CONFIG_TNY_A9263
 	default "0x310000"	if CONFIG_AT91SAM9RLEK
 	default "0x301000"	if CONFIG_AT91SAM9XEEK
 	default "0x301000"	if CONFIG_AT91SAM9G10EK
 	default "0x301000"	if CONFIG_AT91SAM9G20EK
+	default "0x301000"	if CONFIG_SBC35_A9G20
 	default "0x304000"	if CONFIG_AT91SAM9G45EK
 	default "0x304000"	if CONFIG_AT91SAM9G45EKES
 	default "0x301000"	if CONFIG_AT91SAM9M10EK
@@ -332,6 +349,7 @@ config CONFIG_CRYSTAL
 
 config CONFIG_CRYSTAL
 	string
+	default "CRYSTAL_12_000MHZ"	if CONFIG_CRYSTAL_12_000MHZ
 	default "CRYSTAL_16_000MHZ"	if CONFIG_CRYSTAL_16_000MHZ
 	default "CRYSTAL_16_36766MHZ"	if CONFIG_CRYSTAL_16_36766MHZ
 	default "CRYSTAL_18_432MHZ"	if CONFIG_CRYSTAL_18_432MHZ
diff --git a/board/boards b/board/boards
index 9da6dd4..ac82886 100644
--- a/board/boards
+++ b/board/boards
@@ -16,6 +16,7 @@ at91sam9261ek/at91sam9261nf_defconfig
 at91sam9263ek/at91sam9263dfc_defconfig
 at91sam9263ek/at91sam9263ek_defconfig
 at91sam9263ek/at91sam9263nf_defconfig
+tny-a9263/tny-a9263_defconfig
 at91sam9g10ek/at91sam9g10df_defconfig
 at91sam9g10ek/at91sam9g10ek_defconfig
 at91sam9g10ek/at91sam9g10nf_defconfig
diff --git a/board/tny-a9263/board.mk b/board/tny-a9263/board.mk
new file mode 100644
index 0000000..4621397
--- /dev/null
+++ b/board/tny-a9263/board.mk
@@ -0,0 +1,6 @@
+CPPFLAGS += \
+	-DCONFIG_TNY_A9263
+
+ASFLAGS += \
+	-DCONFIG_TNY_A9263
+
diff --git a/board/tny-a9263/tny-a9263.c b/board/tny-a9263/tny-a9263.c
new file mode 100644
index 0000000..31caed1
--- /dev/null
+++ b/board/tny-a9263/tny-a9263.c
@@ -0,0 +1,581 @@
+/* ----------------------------------------------------------------------------
+ *         ATMEL Microcontroller Software Support  -  ROUSSET  -
+ * ----------------------------------------------------------------------------
+ * Copyright (c) 2009, Atmel Corporation
+ * All rights reserved.
+ * PSRAM,16 bit SDRAM support courtesy of Epsilon Group
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the disclaimer below.
+ *
+ * Atmel's name may not be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * DISCLAIMER: THIS SOFTWARE IS PROVIDED BY ATMEL "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT ARE
+ * DISCLAIMED. IN NO EVENT SHALL ATMEL BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ----------------------------------------------------------------------------
+ * File Name           : tny-a9263.c
+ * Object              :
+ * Creation            : GH Dec 21th 2010
+ *-----------------------------------------------------------------------------
+ */
+#include "part.h"
+#include "main.h"
+#include "gpio.h"
+#include "pmc.h"
+#include "rstc.h"
+#include "debug.h"
+#include "memory.h"
+#include "psram.h"
+#include "matrix.h"
+
+#ifndef CONFIG_THUMB
+static inline unsigned int get_cp15(void)
+{
+	unsigned int value;
+	__asm__("mrc p15, 0, %0, c1, c0, 0" : "=r" (value));
+	return value;
+}
+
+static inline void set_cp15(unsigned int value)
+{
+	__asm__("mcr p15, 0, %0, c1, c0, 0" : : "r" (value));
+}
+#else
+int get_cp15(void);
+void set_cp15(unsigned int value);
+#endif
+
+#ifdef CONFIG_HW_INIT
+/*---------------------------------------------------------------------------- */
+/* \fn    hw_init							       */
+/* \brief This function performs very low level HW initialization	       */
+/* This function is invoked as soon as possible during the c_startup	       */
+/* The bss segment must be initialized					       */
+/*---------------------------------------------------------------------------- */
+void hw_init(void)
+{
+	/* Configure PIOs */
+	const struct pio_desc hw_pio[] = {
+#ifdef CONFIG_VERBOSE
+		{"RXD", AT91C_PIN_PC(30), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"TXD", AT91C_PIN_PC(31), 0, PIO_DEFAULT, PIO_PERIPH_A},
+#endif
+		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
+	};
+
+	/* Disable watchdog */
+	writel(AT91C_WDTC_WDDIS, AT91C_BASE_WDTC + WDTC_WDMR);
+
+	/* At this stage the main oscillator is supposed to be enabled
+	 * PCK = MCK = MOSC */
+
+	/* Configure PLLA = MOSC * (PLL_MULA + 1) / PLL_DIVA */
+	pmc_cfg_plla(PLLA_SETTINGS, PLL_LOCK_TIMEOUT);
+
+	/* PCK = PLLA = 2 * MCK */
+	pmc_cfg_mck(MCKR_SETTINGS, PLL_LOCK_TIMEOUT);
+	/* Switch MCK on PLLA output */
+	pmc_cfg_mck(MCKR_CSS_SETTINGS, PLL_LOCK_TIMEOUT);
+
+
+	/* Configure PLLB */
+	pmc_cfg_pllb(PLLB_SETTINGS, PLL_LOCK_TIMEOUT);
+
+	/* Enable External Reset */
+	writel(AT91C_RSTC_KEY_UNLOCK || AT91C_RSTC_URSTEN, AT91C_BASE_RSTC + RSTC_RMR);
+
+	/* Configure the PIO controller to output PCK0 */
+	pio_setup(hw_pio);
+
+	/* Configure TCMs (Tightly coupled memories) */
+
+	/* Configure the EBI0 Slave Slot Cycle to 64 */
+	// writel( (readl((AT91C_BASE_MATRIX + MATRIX_SCFG4)) & ~0xFF) | 0x40, (AT91C_BASE_MATRIX + MATRIX_SCFG4));
+
+	writel(BURST_16WORD_BEAT,	AHB_MATRIX_MASTER_OHCI);
+	writel(BURST_8WORD_BEAT,	AHB_MATRIX_MASTER_ISI);
+	writel(BURST_8WORD_BEAT,	AHB_MATRIX_MASTER_2D);
+	writel(BURST_8WORD_BEAT,	AHB_MATRIX_MASTER_DMAC);
+	writel(BURST_4WORD_BEAT,	AHB_MATRIX_MASTER_MACB);
+	writel(BURST_16WORD_BEAT,	AHB_MATRIX_MASTER_LCDC);
+	writel(BURST_SINGLE_ACCESS,	AHB_MATRIX_MASTER_PDC);
+	writel(BURST_8WORD_BEAT,	AHB_MATRIX_MASTER_DBUS);
+	writel(BURST_4WORD_BEAT,	AHB_MATRIX_MASTER_IBUS);
+
+
+#if	0
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		||
+		AHB_MATRIX_ROUND_ROBIN			|| 
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_ARM926I	||
+		AT91C_MATRIX_FIXED_DEFMSTR0_ARM926D	||
+		AT91C_MATRIX_FIXED_DEFMSTR0_PDC		||
+		AT91C_MATRIX_FIXED_DEFMSTR0_LCDC	||
+		AT91C_MATRIX_FIXED_DEFMSTR0_2DGC	||
+		AT91C_MATRIX_FIXED_DEFMSTR0_ISI		||
+		AT91C_MATRIX_FIXED_DEFMSTR0_DMA		||
+		AT91C_MATRIX_FIXED_DEFMSTR0_EMAC	||
+		AT91C_MATRIX_FIXED_DEFMSTR0_USB		||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_NO_DEFMSTR	||
+		AT91C_MATRIX_DEFMSTR_TYPE_LAST_DEFMSTR	||
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(),
+#endif
+
+	writel(
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		||
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_ARM926I	||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_LAST_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(32),
+		AHB_MATRIX_SLAVE_ROM
+	);
+
+	writel(
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		||
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_EMAC	||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(32),
+		AHB_MATRIX_SLAVE_RAM80K
+	);
+
+	writel(
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		||
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_USB		||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(16),
+		AHB_MATRIX_SLAVE_RAM16K
+	);
+
+	writel(
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		||
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_PDC		||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(4),
+		AHB_MATRIX_SLAVE_PERIPHERALS
+	);
+
+	writel(
+		// ARBT
+		AHB_MATRIX_ROUND_ROBIN			||
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_ARM926I	||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(32),
+		AHB_MATRIX_SLAVE_EBI0
+	);
+
+	writel(
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		|| 
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_LCDC	||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(64),
+		AHB_MATRIX_SLAVE_EBI1
+	);
+
+	writel(
+		// ARBT
+		AHB_MATRIX_FIXED_PRIORITY		||
+
+		// FIXED_DEFMSTR
+		AT91C_MATRIX_FIXED_DEFMSTR0_ARM926D	||
+
+		// DEFMSTR_TYPE
+		AT91C_MATRIX_DEFMSTR_TYPE_FIXED_DEFMSTR	||
+
+		// SLOT_CYCLE
+		SLOT_CYCLE(4),
+		AHB_MATRIX_SLAVE_APB
+	);
+
+	AHB_PRIORITY9(	ABH_PRIORITY_ROM,
+		AHB_IBUS_ID,
+		AHB_DBUS_ID,
+		// Rest should not access the ROM
+		AHB_PDC_ID,
+		AHB_DMAC_ID,
+		AHB_OHCI_ID,
+		AHB_ISI_ID,
+		AHB_2D_ID,
+		AHB_MACB_ID,
+		AHB_LCDC_ID
+	);
+	AHB_PRIORITY9(	ABH_PRIORITY_RAM80K,
+		AHB_PDC_ID,
+		AHB_MACB_ID,
+		AHB_DMAC_ID,
+		AHB_OHCI_ID,
+		AHB_IBUS_ID,
+		AHB_DBUS_ID,
+		// Rest should not access the large SRAM
+		AHB_ISI_ID,
+		AHB_2D_ID,
+		AHB_LCDC_ID
+	);
+	AHB_PRIORITY9(	ABH_PRIORITY_RAM16K,
+		AHB_OHCI_ID,
+		AHB_PDC_ID,
+		AHB_DMAC_ID,
+		AHB_DBUS_ID,
+		AHB_IBUS_ID,
+		// Rest should not access the small SRAM
+		AHB_ISI_ID,
+		AHB_2D_ID,
+		AHB_MACB_ID,
+		AHB_LCDC_ID
+	);
+	AHB_PRIORITY9(	ABH_PRIORITY_PERIPHERALS,
+		AHB_DBUS_ID,
+		// Rest should not access & reconfigure peripherals
+		AHB_PDC_ID,
+		AHB_DMAC_ID,
+		AHB_OHCI_ID,
+		AHB_ISI_ID,
+		AHB_2D_ID,
+		AHB_MACB_ID,
+		AHB_LCDC_ID,
+		AHB_IBUS_ID
+	);
+
+#if	defined(CONFIG_PSRAM)
+	// LCD on EBI1
+	AHB_PRIORITY9(	ABH_PRIORITY_EBI0,
+		AHB_PDC_ID,
+		AHB_MACB_ID,
+		AHB_OHCI_ID,
+		AHB_DMAC_ID,
+		AHB_IBUS_ID,
+		AHB_DBUS_ID,
+		// Rest should not access EBI0
+		AHB_ISI_ID,
+		AHB_LCDC_ID,
+		AHB_2D_ID	
+	);
+#else
+	// LCD on EBI0
+	AHB_PRIORITY9(	ABH_PRIORITY_EBI0,
+		AHB_PDC_ID,
+		AHB_LCDC_ID,
+		AHB_MACB_ID,
+		AHB_OHCI_ID,
+		AHB_DMAC_ID,
+		AHB_2D_ID,
+		AHB_IBUS_ID,
+		AHB_DBUS_ID,
+		// Rest should not access EBI0
+		AHB_ISI_ID
+	);
+#endif
+
+	AHB_PRIORITY9(	ABH_PRIORITY_EBI1,
+		AHB_LCDC_ID,
+		AHB_PDC_ID,
+		AHB_DMAC_ID,
+		AHB_2D_ID,
+		AHB_DBUS_ID,
+		// Rest should not access ABI1
+		AHB_IBUS_ID,
+		AHB_OHCI_ID,
+		AHB_ISI_ID,
+		AHB_MACB_ID
+	);
+
+	AHB_PRIORITY9(	ABH_PRIORITY_APB,
+		AHB_PDC_ID,
+		AHB_DMAC_ID,
+		AHB_DBUS_ID,
+		// Rest should not access APB
+		AHB_IBUS_ID,
+		AHB_OHCI_ID,
+		AHB_ISI_ID,
+		AHB_2D_ID,
+		AHB_MACB_ID,
+		AHB_LCDC_ID
+	);
+
+#ifdef CONFIG_VERBOSE
+	/* Enable Debug messages on the DBGU */
+	dbg_init(BAUDRATE(MASTER_CLOCK, 115200));
+	header();
+#endif /* CONFIG_DEBUG */
+
+#ifdef CONFIG_SDRAM
+	/* Initialize the matrix */
+	/* VDDIOMSEL = 1 -> Memories are 3.3V powered */
+	writel(readl(AT91C_BASE_CCFG + CCFG_EBI0CSA) | (1 << 16) | AT91C_EBI_CS1A_SDRAMC, AT91C_BASE_CCFG + CCFG_EBI0CSA);
+	/* Configure SDRAM Controller */
+#if	defined(CONFIG_SDRAM_16BIT)
+	sdram_init(	AT91C_SDRAMC_NC_10  |
+				AT91C_SDRAMC_NR_13 |
+				AT91C_SDRAMC_CAS_2 |
+				AT91C_SDRAMC_NB_4_BANKS |
+				AT91C_SDRAMC_DBW_16_BITS |
+				AT91C_SDRAMC_TWR_2 |
+				AT91C_SDRAMC_TRC_7 |
+				AT91C_SDRAMC_TRP_2 |
+				AT91C_SDRAMC_TRCD_2 |
+				AT91C_SDRAMC_TRAS_5 |
+				AT91C_SDRAMC_TXSR_8,		/* Control Register       */
+				(MASTER_CLOCK * 7)/1000000,	/* Refresh Timer Register */
+				AT91C_SDRAMC_MD_SDRAM);		/* SDRAM (no low power)   */
+#else
+	sdram_init(	AT91C_SDRAMC_NC_9  |
+				AT91C_SDRAMC_NR_13 |
+				AT91C_SDRAMC_CAS_2 |
+				AT91C_SDRAMC_NB_4_BANKS |
+				AT91C_SDRAMC_DBW_32_BITS |
+				AT91C_SDRAMC_TWR_2 |
+				AT91C_SDRAMC_TRC_7 |
+				AT91C_SDRAMC_TRP_2 |
+				AT91C_SDRAMC_TRCD_2 |
+				AT91C_SDRAMC_TRAS_5 |
+				AT91C_SDRAMC_TXSR_8,    /* Control Register       */
+				(MASTER_CLOCK * 7)/1000000, /* Refresh Timer Register */
+				AT91C_SDRAMC_MD_SDRAM);   /* SDRAM (no low power)   */
+#endif
+#endif	/* CONFIG_SDRAM */
+
+
+#if	defined(CONFIG_PSRAM)
+	 psram_hw_init();
+#endif  /* CFG_PSRAM */
+}
+#endif /* CONFIG_HW_INIT */
+
+
+#ifdef CONFIG_SDRAM
+//*----------------------------------------------------------------------------
+//* \fn    sdramc_hw_init
+//* \brief This function performs SDRAMC HW initialization
+//*----------------------------------------------------------------------------*/
+void sdramc_hw_init(void)
+{
+	/* Configure PIOs */
+	const struct pio_desc sdramc_pio[] = {
+		{"D16", AT91C_PIN_PD(16), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D17", AT91C_PIN_PD(17), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D18", AT91C_PIN_PD(18), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D19", AT91C_PIN_PD(19), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D20", AT91C_PIN_PD(20), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D21", AT91C_PIN_PD(21), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D22", AT91C_PIN_PD(22), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D23", AT91C_PIN_PD(23), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D24", AT91C_PIN_PD(24), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D25", AT91C_PIN_PD(25), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D26", AT91C_PIN_PD(26), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D27", AT91C_PIN_PD(27), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D28", AT91C_PIN_PD(28), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D29", AT91C_PIN_PD(29), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D30", AT91C_PIN_PD(30), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{"D31", AT91C_PIN_PD(31), 0, PIO_DEFAULT, PIO_PERIPH_A},
+		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
+	};
+
+	/* Configure the SDRAMC PIO controller */
+	pio_setup(sdramc_pio);
+}
+#endif
+
+#ifdef CONFIG_PSRAM
+void psram_hw_init(void)
+{
+	volatile unsigned short tmp;
+	unsigned short* addressMax = (unsigned short*)MICRON_8MB_ADDRESS_MAX;
+	/* Configure PIOs */
+	const struct pio_desc psram_pio[] = {
+		{"CRE", AT91C_PIN_PE(16), 0, PIO_DEFAULT, PIO_OUTPUT},
+		{(char *) 0, 0, 0,  PIO_DEFAULT, PIO_PERIPH_A},
+	};
+
+	writel((AT91C_PSRAM_NWE_SETUP |
+		AT91C_PSRAM_NCS_WR_SETUP |
+		AT91C_PSRAM_NRD_SETUP |
+		AT91C_PSRAM_NCS_RD_SETUP),
+		AT91C_BASE_SMC1 + SMC_SETUP0);
+	writel((AT91C_PSRAM_NWE_PULSE |
+		AT91C_PSRAM_NCS_WR_PULSE |
+		AT91C_PSRAM_NRD_PULSE |
+		AT91C_PSRAM_NCS_RD_PULSE),
+		AT91C_BASE_SMC1 + SMC_PULSE0);
+	writel((AT91C_PSRAM_NWE_CYCLE |
+		AT91C_PSRAM_NRD_CYCLE),
+		AT91C_BASE_SMC1 + SMC_CYCLE0);
+	writel((AT91C_SMC_READMODE |
+		AT91C_SMC_WRITEMODE |
+		AT91C_SMC_BAT_BYTE_SELECT |
+		AT91C_SMC_PMEN |
+		AT91C_SMC_DBW_WIDTH_SIXTEEN_BITS |
+		AT91C_SMC_PS_SIZE_THIRTY_TWO_BYTES),
+		AT91C_BASE_SMC1 + SMC_CTRL0);
+
+	pio_setup(psram_pio);
+	pio_set_value(AT91C_PIN_PE(16), 0);  // Data access.
+
+	//Enable page mode
+	tmp = readl(addressMax);
+	tmp = readl(addressMax);
+	writel(MICRON_RCR,addressMax);
+	writel(MICRON_PAGE_MODE_ENABLE,addressMax);
+}
+#endif  /* CONFIG_PSRAM */
+
+#ifdef CONFIG_DATAFLASH
+#if	defined(CONFIG_DATAFLASH_RECOVERY)
+/*------------------------------------------------------------------------------*/
+/* \fn    df_recovery								*/
+/* \brief This function erases DataFlash Page 0 if USER PB is pressed 		*/
+/*        during boot sequence							*/
+/*------------------------------------------------------------------------------*/
+void df_recovery(AT91PS_DF pDf)
+{
+#if (AT91C_SPI_PCS_DATAFLASH == AT91C_SPI_PCS0_DATAFLASH)
+	/* Configure PIOs */
+	const struct pio_desc userbp_pio[] = {
+		{"USERPB", AT91C_PIN_PB(10), 0, PIO_PULLUP, PIO_INPUT},
+		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
+	};
+
+	/* Configure the PIO controller */
+	writel((1 << AT91C_ID_PIOB), PMC_PCER + AT91C_BASE_PMC);
+	pio_setup(userbp_pio);
+	
+	/* If USER PB is pressed during Boot sequence */
+	/* Erase DateFlash page 0*/
+	if ( !pio_get_value(AT91C_PIN_PB(10)) )
+		df_page_erase(pDf, 0);
+#endif
+}
+#endif
+#endif
+
+#ifdef CONFIG_DATAFLASH
+/*------------------------------------------------------------------------------*/
+/* \fn    df_hw_init								*/
+/* \brief This function performs DataFlash HW initialization			*/
+/*------------------------------------------------------------------------------*/
+void df_hw_init(void)
+{
+	/* Configure PIOs */
+	const struct pio_desc df_pio[] = {
+		{"MISO",  AT91C_PIN_PA(0), 0, PIO_DEFAULT, PIO_PERIPH_B},
+		{"MOSI",  AT91C_PIN_PA(1), 0, PIO_DEFAULT, PIO_PERIPH_B},
+		{"SPCK",  AT91C_PIN_PA(2), 0, PIO_DEFAULT, PIO_PERIPH_B},
+		{"NPCS0", AT91C_PIN_PA(5), 0, PIO_DEFAULT, PIO_PERIPH_B},
+		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
+	};
+
+	/* Configure the PIO controller */
+	pio_setup(df_pio);
+}
+#endif /* CONFIG_DATAFLASH */
+
+
+#ifdef CONFIG_NANDFLASH
+/*------------------------------------------------------------------------------*/
+/* \fn    nandflash_hw_init							*/
+/* \brief NandFlash HW init							*/
+/*------------------------------------------------------------------------------*/
+void nandflash_hw_init(void)
+{
+	/* Configure PIOs */
+	const struct pio_desc nand_pio[] = {
+		{"RDY_BSY", AT91C_PIN_PA(22), 0, PIO_PULLUP, PIO_INPUT},
+		{"NANDCS",  AT91C_PIN_PD(15), 0, PIO_PULLUP, PIO_OUTPUT},
+		{(char *) 0, 0, 0,  PIO_DEFAULT, PIO_PERIPH_A},
+	};
+
+	/* Setup Smart Media, first enable the address range of CS3 in HMATRIX user interface */
+	writel(readl(AT91C_BASE_CCFG + CCFG_EBI0CSA) | AT91C_EBI_CS3A_SM, AT91C_BASE_CCFG + CCFG_EBI0CSA);
+		    
+	/* Configure SMC CS3 */
+ 	writel((AT91C_SM_NWE_SETUP | AT91C_SM_NCS_WR_SETUP | AT91C_SM_NRD_SETUP | AT91C_SM_NCS_RD_SETUP), AT91C_BASE_SMC0 + SMC_SETUP3);
+  	writel((AT91C_SM_NWE_PULSE | AT91C_SM_NCS_WR_PULSE | AT91C_SM_NRD_PULSE | AT91C_SM_NCS_RD_PULSE), AT91C_BASE_SMC0 + SMC_PULSE3);
+	writel((AT91C_SM_NWE_CYCLE | AT91C_SM_NRD_CYCLE)						, AT91C_BASE_SMC0 + SMC_CYCLE3);
+	writel((AT91C_SMC_READMODE | AT91C_SMC_WRITEMODE | AT91C_SMC_NWAITM_NWAIT_DISABLE | 
+  		AT91C_SMC_DBW_WIDTH_SIXTEEN_BITS | AT91C_SM_TDF)						, AT91C_BASE_SMC0 + SMC_CTRL3);
+
+	/* Configure the PIO controller */
+	writel((1 << AT91C_ID_PIOA), PMC_PCER + AT91C_BASE_PMC);
+	writel((1 << AT91C_ID_PIOCDE), PMC_PCER + AT91C_BASE_PMC);
+
+	pio_setup(nand_pio);
+
+}
+
+/*------------------------------------------------------------------------------*/
+/* \fn    nandflash_cfg_16bits_dbw_init						*/
+/* \brief Configure SMC in 16 bits mode						*/
+/*------------------------------------------------------------------------------*/
+void nandflash_cfg_16bits_dbw_init(void)
+{
+	writel(readl(AT91C_BASE_SMC0 + SMC_CTRL3) | AT91C_SMC_DBW_WIDTH_SIXTEEN_BITS, AT91C_BASE_SMC0 + SMC_CTRL3);
+}
+
+/*------------------------------------------------------------------------------*/
+/* \fn    nandflash_cfg_8bits_dbw_init						*/
+/* \brief Configure SMC in 8 bits mode						*/
+/*------------------------------------------------------------------------------*/
+void nandflash_cfg_8bits_dbw_init(void)
+{
+	writel((readl(AT91C_BASE_SMC0 + SMC_CTRL3) & ~(AT91C_SMC_DBW)) | AT91C_SMC_DBW_WIDTH_EIGTH_BITS, AT91C_BASE_SMC0 + SMC_CTRL3);
+}
+
+#endif /* #ifdef CONFIG_NANDFLASH */
+
+
diff --git a/board/tny-a9263/tny-a9263.h b/board/tny-a9263/tny-a9263.h
new file mode 100644
index 0000000..33beb02
--- /dev/null
+++ b/board/tny-a9263/tny-a9263.h
@@ -0,0 +1,137 @@
+/* ----------------------------------------------------------------------------
+ *         ATMEL Microcontroller Software Support  -  ROUSSET  -
+ * ----------------------------------------------------------------------------
+ * Copyright (c) 2006, Atmel Corporation
+
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the disclaimer below.
+ *
+ * Atmel's name may not be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * DISCLAIMER: THIS SOFTWARE IS PROVIDED BY ATMEL "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT ARE
+ * DISCLAIMED. IN NO EVENT SHALL ATMEL BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ----------------------------------------------------------------------------
+ * File Name           : tny-a9263.h
+ * Object              :
+ * Creation            : GH Dec 21th 2010
+ *-----------------------------------------------------------------------------
+ */
+#ifndef _TNY_A9263_H
+#define _TNY_A9263_H
+
+/* ******************************************************************* */
+/* PMC Settings                                                        */
+/*                                                                     */
+/* The main oscillator is enabled as soon as possible in the c_startup */
+/* and MCK is switched on the main oscillator.                         */
+/* PLL initialization is done later in the hw_init() function          */
+/* ******************************************************************* */
+#if	defined(CONFIG_CPU_CLK_180MHZ)
+
+	#define MASTER_CLOCK		(180000000/2)
+	#define PLL_LOCK_TIMEOUT	1000000
+	#define PLLA_SETTINGS	0x20593F06
+	#define PLLB_SETTINGS	0x10483F0E
+
+#endif
+
+#if	defined(CONFIG_CPU_CLK_200MHZ)
+
+	#define MASTER_CLOCK		(200000000/2)
+	#define PLL_LOCK_TIMEOUT	1000000
+	#define PLLA_SETTINGS	0x20633F06
+	#define PLLB_SETTINGS	0x10483F0E
+
+#endif
+
+#if	defined(CONFIG_CPU_CLK_240MHZ)
+
+	#define MASTER_CLOCK		(240000000/2)
+	#define PLL_LOCK_TIMEOUT	1000000
+	#define PLLA_SETTINGS		0x20133F01
+	#define PLLB_SETTINGS		0x10483F0E
+#endif
+
+
+#define	TOP_OF_MEMORY		0x314000
+
+/* Switch MCK on PLLA output PCK = PLLA = 2 * MCK */
+#define MCKR_SETTINGS		(AT91C_PMC_PRES_CLK | AT91C_PMC_MDIV_2)
+#define MCKR_CSS_SETTINGS	(AT91C_PMC_CSS_PLLA_CLK | MCKR_SETTINGS)
+
+/* ******************************************************************* */
+/* DataFlash Settings                                                  */
+/*                                                                     */
+/* ******************************************************************* */
+#define AT91C_BASE_SPI	AT91C_BASE_SPI0
+#define AT91C_ID_SPI	AT91C_ID_SPI0
+
+/* AC characteristics */
+/* DLYBS = tCSS= 250ns min and DLYBCT = tCSH = 250ns */
+#define DATAFLASH_TCSS		(0x1a << 16)	/* 250ns min (tCSS) <=> 12/48000000 = 250ns */
+#define DATAFLASH_TCHS		(0x1 << 24)	/* 250ns min (tCSH) <=> (64*1+SCBR)/(2*48000000) */
+
+#define DF_CS_SETTINGS 		(AT91C_SPI_NCPHA | (AT91C_SPI_DLYBS & DATAFLASH_TCSS) | (AT91C_SPI_DLYBCT & DATAFLASH_TCHS) | ((MASTER_CLOCK / AT91C_SPI_CLK) << 8))
+
+/* ******************************************************************* */
+/* SDRAMC Settings                                                     */
+/*                                                                     */
+/* ******************************************************************* */
+#define AT91C_BASE_SDRAMC 	AT91C_BASE_SDRAMC0
+#define AT91C_EBI_SDRAM		AT91C_EBI0_SDRAM
+
+/* ******************************************************************* */
+/* NandFlash Settings                                                  */
+/*                                                                     */
+/* ******************************************************************* */
+#define AT91C_SMARTMEDIA_BASE	0x40000000
+
+#define AT91_SMART_MEDIA_ALE    (1 << 21)	/* our ALE is AD21 */
+#define AT91_SMART_MEDIA_CLE    (1 << 22)	/* our CLE is AD22 */
+
+#define NAND_DISABLE_CE() do { *(volatile unsigned int *)AT91C_PIOD_SODR = AT91C_PIO_PD15;} while(0)
+#define NAND_ENABLE_CE()  do { *(volatile unsigned int *)AT91C_PIOD_CODR = AT91C_PIO_PD15;} while(0)
+#define NAND_WAIT_READY() while (!(*(volatile unsigned int *)AT91C_PIOA_PDSR & AT91C_PIO_PA22))
+
+/* ******************************************************************* */
+/* SDRAMC Settings                                                     */
+/*                                                                     */
+/* ******************************************************************* */
+#define AT91C_BASE_SDRAMC 	AT91C_BASE_SDRAMC0
+#define AT91C_EBI_SDRAM		AT91C_EBI0_SDRAM
+
+/* ******************************************************************** */
+/* SMC Chip Select 3 Timings for NandFlash for MASTER_CLOCK = 100000000.*/
+/* Please refer to SMC section in AT91SAM9x datasheet to learn how 	*/
+/* to generate these values. 						*/
+/* ******************************************************************** */
+#define AT91C_SM_NWE_SETUP	(1 << 0)
+#define AT91C_SM_NCS_WR_SETUP	(1 << 8)
+#define AT91C_SM_NRD_SETUP	(1 << 16)
+#define AT91C_SM_NCS_RD_SETUP	(1 << 24)
+  
+#define AT91C_SM_NWE_PULSE 	(3 << 0)
+#define AT91C_SM_NCS_WR_PULSE	(3 << 8)
+#define AT91C_SM_NRD_PULSE	(3 << 16)
+#define AT91C_SM_NCS_RD_PULSE	(3 << 24)
+  
+#define AT91C_SM_NWE_CYCLE 	(5 << 0)
+#define AT91C_SM_NRD_CYCLE	(5 << 16)
+
+#define AT91C_SM_TDF	        (2 << 16)
+
+#endif	/* _TNY_A9263_H */
diff --git a/board/tny-a9263/tny-a9263nf_defconfig b/board/tny-a9263/tny-a9263nf_defconfig
new file mode 100644
index 0000000..71035c0
--- /dev/null
+++ b/board/tny-a9263/tny-a9263nf_defconfig
@@ -0,0 +1,90 @@
+#
+# Automatically generated make config: don't edit
+# Tue Dec 21 14:49:53 2010
+#
+HAVE_DOT_CONFIG=y
+CONFIG_BOARDNAME="tny-a9263"
+# CONFIG_AT91SAM9260EK is not set
+# CONFIG_AT91SAM9261EK is not set
+# CONFIG_AT91SAM9263EK is not set
+CONFIG_TNY_A9263=y
+# CONFIG_AT91SAM9RLEK is not set
+# CONFIG_AT91SAM9XEEK is not set
+# CONFIG_AT91SAM9G10EK is not set
+# CONFIG_AT91SAM9G20EK is not set
+# CONFIG_AT91SAM9G45EK is not set
+# CONFIG_AT91SAM9G45EKES is not set
+# CONFIG_AT91SAM9M10EK is not set
+# CONFIG_AT91CAP9ADK is not set
+# CONFIG_AT91CAP9STK is not set
+# CONFIG_AFEB9260 is not set
+CONFIG_CHIP="AT91SAM9263"
+CONFIG_BOARD="tny-a9263"
+CONFIG_MACH_TYPE="0x85C"
+CONFIG_LINK_ADDR="0x000000"
+CONFIG_TOP_OF_MEMORY="0x314000"
+CONFIG_BOOTSTRAP_MAXSIZE="13000"
+CONFIG_CRYSTAL_12_000MHZ=y
+# CONFIG_CRYSTAL_16_000MHZ is not set
+# CONFIG_CRYSTAL_16_36766MHZ is not set
+# CONFIG_CRYSTAL_18_432MHZ is not set
+ALLOW_CRYSTAL_12_000MHZ=y
+CONFIG_CRYSTAL="CRYSTAL_12_000MHZ"
+# CONFIG_CPU_CLK_166MHZ is not set
+CONFIG_CPU_CLK_180MHZ=y
+# CONFIG_CPU_CLK_200MHZ is not set
+# CONFIG_CPU_CLK_240MHZ is not set
+# CONFIG_CPU_CLK_266MHZ is not set
+# CONFIG_CPU_CLK_400MHZ is not set
+ALLOW_CPU_CLK_180MHZ=y
+ALLOW_CPU_CLK_200MHZ=y
+ALLOW_CPU_CLK_240MHZ=y
+# DISABLE_CPU_CLK_240MHZ is not set
+# CONFIG_BUS_SPEED_83MHZ is not set
+CONFIG_BUS_SPEED_90MHZ=y
+# CONFIG_BUS_SPEED_100MHZ is not set
+# CONFIG_BUS_SPEED_133MHZ is not set
+
+#
+# Memory selection
+#
+CONFIG_SDRAM=y
+# CONFIG_SDDRC is not set
+# CONFIG_DDR2 is not set
+# ALLOW_DATAFLASH is not set
+# ALLOW_FLASH is not set
+ALLOW_NANDFLASH=y
+# ALLOW_SDCARD is not set
+# ALLOW_PSRAM is not set
+# ALLOW_SDRAM_16BIT is not set
+# CONFIG_RAM_32MB is not set
+CONFIG_RAM_64MB=y
+# CONFIG_RAM_128MB is not set
+# CONFIG_RAM_256MB is not set
+# CONFIG_DATAFLASH is not set
+# CONFIG_FLASH is not set
+CONFIG_NANDFLASH=y
+# CONFIG_SDCARD is not set
+CONFIG_MEMORY="nandflash"
+
+#
+# NAND Flash configuration
+#
+# CONFIG_NANDFLASH_SMALL_BLOCKS is not set
+CONFIG_PROJECT="nandflash"
+CONFIG_LOAD_UBOOT=y
+# CONFIG_LOAD_64KB is not set
+# CONFIG_LOAD_1MB is not set
+# CONFIG_LOAD_4MB is not set
+CONFIG_IMG_ADDRESS="0x00020000"
+CONFIG_IMG_SIZE="0x00040000"
+CONFIG_JUMP_ADDR="0x23F00000"
+CONFIG_APP_CHECK=y
+# CONFIG_LONG_TEST is not set
+# CONFIG_DEBUG is not set
+# CONFIG_VERBOSE is not set
+CONFIG_HW_INIT=y
+# CONFIG_USER_HW_INIT is not set
+# CONFIG_THUMB is not set
+CONFIG_SX_AT91=y
+CONFIG_RAW_AT91=y
diff --git a/driver/Config.in.dataflash b/driver/Config.in.dataflash
index 445e053..8c9ab06 100644
--- a/driver/Config.in.dataflash
+++ b/driver/Config.in.dataflash
@@ -103,7 +103,7 @@ config CONFIG_DATAFLASHCARD
 
 config CONFIG_CARD_SUFFIX
 	string
-	default "card" if CONFIG_DATAFLASHCARD
+	default "" if CONFIG_DATAFLASHCARD
 	default ""
 
 config CONFIG_SPI_BOOT
diff --git a/include/part.h b/include/part.h
index 3aedd6a..48ce999 100644
--- a/include/part.h
+++ b/include/part.h
@@ -60,8 +60,10 @@
 
 #ifdef AT91SAM9G20
 #include "AT91SAM9260_inc.h"
+#ifdef CONFIG_AT91SAM9G20EK
 #include "at91sam9g20ek.h"
 #endif
+#endif
 
 #ifdef AT91SAM9G45
 #include "AT91SAM9G45_inc.h"
@@ -85,8 +87,13 @@
 
 #ifdef AT91SAM9263
 #include "AT91SAM9263_inc.h"
+#ifdef CONFIG_AT91SAM9263EK
 #include "at91sam9263ek.h"
 #endif
+#ifdef CONFIG_TNY_A9263
+#include "tny-a9263.h"
+#endif
+#endif
 
 #ifdef AT91CAP9
 #include "AT91CAP9_inc.h"
